### ISSUE TITLE
.first_or_create()

### ISSUE THREAD

**Issue by ckhatton** [dominicsayers/mailchimp_api_v3#13]
I have found an issue with the `first_or_create` function. If the record doesn't exist, it is created. If the record does exist, it does not return it, but instead throws a `Mailchimp::Exception::BadRequest`. Is that what should happen? 🐙

**Comment by dominicsayers** [comment]
It seems OK for me. I tried something like this:

```
mailchimp = Mailchimp.connect
mailchimp.lists[0].members.first_or_create name: 'Dominic Sayers'
```

Internally, `first_or_create` calls the `find_by` method. Can you repeat your test using `find_by` and let me know if this also throws an exception?

Could you also share some code to give more details of how you were using the `first_or_create` method?

**Comment by ckhatton** [comment]
Right, back to this now. I was wrong, it throws a `Mailchimp::Exception::Duplicate`, which is expected, but ideally a record should be returned right? I see the `return` line [here](https://github.com/dominicsayers/mailchimp_api_v3/blob/develop/lib/mailchimp_api_v3/collection.rb#L47).

If I use `find_by` there are no errors, but nothing is returned. :suspect:

```ruby
@emails.each do |email|
  attributes = {
    "email_address": email,
    "status": "subscribed"
  }

  record = MAILCHIMP.lists('0719f1b43e').members.find_by(attributes)
  puts record
end
```

**Comment by ckhatton** [comment]
I have previously gotten around it by manually going through the members list and checking if a record exists. The record is updated if it does, otherwise a new record is created...

```ruby
update_member

  memberFound = false

  list = MAILCHIMP.lists('0719f1b43e')

  list.members.find_each do |member|
    if member.email_address == @email
      # Update the merge fields
      member.update merge_fields: member_attributes[:merge_fields]

      # Update the interest group
      member.update interests: member_attributes[:interests]

      memberFound = true
    end
  end

  if memberFound == false
    MAILCHIMP.lists('0719f1b43e').members.first_or_create(member_attributes)
  end

end
```

**Comment by ckhatton** [comment]
Not sure if it helps, but I also get a fatal error if I use it in an 'each' block where the member **already exists** in the MailChimp list.

```
NoMethodError (undefined method `casecmp' for {"e6037a062c"=>true, "a25411f7a5"=>false}:Hash):
  
app/services/mailchimp_members_service.rb:63:in `block in refer_members'
```

Taking into account that...

```
@emails = ["mail@ckhatton.com"]
ENV['MAILCHIMP_INTEREST_1ST'] = 'e6037a062c'
ENV['MAILCHIMP_INTEREST_2ND'] = 'a25411f7a5'
```

This is my code...

```ruby
def refer_members

  @emails.each do |email|
    attributes = {
      "email_address": email,
      "status": "pending",
      "interests": {
        ENV['MAILCHIMP_INTEREST_1ST'] => true,
        ENV['MAILCHIMP_INTEREST_2ND'] => false
      }
    }

    MAILCHIMP.lists(ENV['MAILCHIMP_LIST_ID']).members.first_or_create(attributes)
  end

end
```

🦑

**Comment by ckhatton** [comment]
I think I may know why nothing is returned...

[#L32](https://github.com/dominicsayers/mailchimp_api_v3/blob/develop/lib/mailchimp_api_v3/collection.rb#L32): Shouldn't that be `return instances`?

[#L37](https://github.com/dominicsayers/mailchimp_api_v3/blob/develop/lib/mailchimp_api_v3/collection.rb#L37): Shouldn't that be `instances ? return instances.first : return nil`?

I can't remember in ruby if stating a variable alone (like [#L32](https://github.com/dominicsayers/mailchimp_api_v3/blob/develop/lib/mailchimp_api_v3/collection.rb#L32)) automatically returns the value.

**Comment by dominicsayers** [comment]
No, Ruby returns the last thing that was evaluated in the method. It's normal Ruby style to not use `return` unless you want to return early from a method.

**Comment by dominicsayers** [comment]
On the more substantive point, I copied your code fragment into a script and ran it:

```Ruby
MAILCHIMP = Mailchimp.connect
@emails = ['ann@sayers.cc']

@emails.each do |email|
  attributes = {
    "email_address": email,
    "status": "subscribed"
  }

  record = MAILCHIMP.lists('e0fd0bddcb').members.find_by(attributes)
  puts record
end
```

and it output

```
<ann@sayers.cc>
```

as expected.

**Comment by ckhatton** [comment]
Arr I think I know why... What does the angle brackets mean? As it outputs as nothing ('nil') in the logs. I know there is something there because I have tried `puts record.nil?`, which returns 'false'.

If I run `puts record` in the rails console, I get...

```
<mail@ckhatton.com>
 => nil
```

If I run `puts 'Account: ' + record` in the rails console, I get...

`TypeError: no implicit conversion of Mailchimp::List::Member into String`

As the result is 'nil', I think that is why I don't see the output in the log.

> The same happens for both `find_by` and `first_or_create`

**Comment by dominicsayers** [comment]
```
puts record
```

explicitly invokes the `to_s` method of `Mailchimp::List:Member` (https://github.com/dominicsayers/mailchimp_api_v3/blob/develop/lib/mailchimp_api_v3/member.rb#L50)

`'Account: ' + record` might imply a different sort of addition and Ruby complains in case you meant something different.

**Comment by ckhatton** [comment]
Ar okay. So how can I get the log to output the string instead of the result 'nil'?
